### PR TITLE
11535 soft launch modal

### LINKED ISSUE #11535: Redesigned homepage - soft launch modal
## Issue description
As a Veteran, I want to be given the opportunity to check out the redesigned homepage once per session so that I can know what's coming and offer feedback.

As the Sitewide Homepage team, we need a way to offer the opportunity to use the new redesigned homepage experience to Veteran and beneficiaries visiting VA.gov. so that we can gather feedback to determine readiness for cutover. 

### Technical considerations

- Ensure the modal **only** displays on VA.gov homepage (`/`)
- Ensure the modal doesn't display on injected sites
- Ensure our modal doesn't display if another modal is called on page load
  - The modal we know about is [modal that appears for users redirected from explore.va.gov](https://www.va.gov/?from=explore.va.gov)

<details>
<summary> Background and pros/cons for banners and modals </summary>

A similar need occurred during the original launch of VA.gov but was managed using a preview subdomain which is not an option for this launch. 

Two options were evaluated during a design, accessibility and implementation discussion
- a banner, allowing a user to move between the existing and new experiences 
- a pop-up modal allowing a user to "opt in" to the new experience. 

**The modal was determined to be the preferred approach.** 

Banners
 - (+) avoid the risk of stacking modals
 - (-) could unintentionally introduce user distrust 
 - (-) would create an escape route preventing data and feedback collection
 - (+/-) banners are less prominent 
 - (-) banner blindness
Modal
 - (-) stacking risk presents usability and accessibility concerns 
 - (+) prominent display
 - (+) making it dismissible will manage annoyance factor and various user flows
</details>

## Expected behavior
- The redesigned experience modal is presented **only** on the VA.gov homepage.
- A Veteran or beneficiary can interact with the modal to agree or decline to use the redesigned homepage.
- The modal offering the redesigned homepage experience displays **only once** within a user's session.
- Regardless of previous opt-in or decline, the modal will be presented during the user's next session (which includes the homepage).
- If the Veteran or beneficiary opts-in to the  redesigned homepage experience, the redesigned page opens in the same window and tab.
- The Veteran or beneficiary can return to the current homepage by using the VA logo and/or back button. The breadcrumb may also be an option, depending on the page. 

_Acceptable known risk:
Veterans or beneficiaries who opt-in to the redesigned experience may not be able to easily return to the redesigned page after navigating to deeper content pages. The user_ **may** _be able to return to the redesigned page using the back button. This user will have another opportunity to opt-in from the homepage during the next VA.gov session._

## Pending
- [x] _Final_ Design from Ryan _(expected by 11/11)_
- [ ] _Final_ Content input _(draft content now available)_
- [ ] Resolution for stacking modals

## Example Modal 

Reference Design system modal: https://design.va.gov/components/modal

## Design

image (19)

* [Slack thread](https://dsva.slack.com/archives/C01DBGX4P45/p1669152245109399) with design system about approved ways to handle the 2 CTA links
  * NOTE: use design system components as-is, no special stying should be required. Mockup above was done in Mural so alignments, padding, sizing, etc is not accurate
* ~~[Sketch design](https://www.sketch.com/s/3aa40506-4be2-46cc-876b-93f1a9f3a857/a/uuid/EEA4DD2D-B2EA-42D3-B10D-DC5A3EA958B7) (NOT FINAL as of 11/23 -- Sketchwas not caught up with the design system discussion above, so rely on the image here / slack thread for official guidance)~~ _outdated for the modal design_

## Draft content

> [h1] Try our new VA.gov homepage
> 
> We're redesigning the VA.gov homepage to help you get the tools and information you need faster. 

> And we want your feedback to help us make it even better. 
> 
> [cta1] Try the new homepage [semantic link, Primary action link from design system]
> [cta2] Not today, go to the current homepage [semantic button, secondary button from design system] 

## Tasks
- Implement redesigned experience modal using cookies to govern display within a user's session
- Configure modal to report analytics
- Focus management and experience via keyboard, voice or screenreader should follow patterns established by other modals in use across VA.gov. 

## Acceptance criteria
- [x] Modal must sit behind a Flipper for both Staging and Prod -  #11653 tracks creation of the Flipper
- [x] When a Veteran or beneficiary arrives on the VA.gov homepage, a modal to opt-in to the redesigned homepage experience overlays the existing homepage.
- [x] Selecting to opt-in with [cta1] link "Try the new homepage" opens the redesigned homepage URL (`/new-home-page`) in the same window/tab. 
- [x] Selecting to opt-out with [cta2] button "Not today, go to the current homepage" closes the modal, displaying the existing/current homepage URL in the same window/tab. 
- [x] Selecting the "x" will close the modal and the existing/current homepage URL will be displayed in the same window/tab. 
- [x] Pressing the ESC key will close the modal and the existing/current homepage URL will be displayed in the same window/tab. 
- [x] The redesigned experience modal _**only**_ displays VA.gov homepage (`/`)
- [ ] The redesigned experience modal is displayed only once during a user's session, regardless how many times the user returns to the homepage. 
- [ ] ~~This modal IS NOT DISPLAYED if another modal should also be present (e.g., the Explore VA redirect modal)~~
- [x] This modal IS NOT DISPLAYED on injected-header/footer sites

**Analytics**
  - [x] When a user opts in to the new homepage, the modal reports analytics with  `event: new-homepage-modal-click` + `modal-primaryButton-text: try the new homepage`
  - [x] When a user declines the new homepage, the modal reports analytics with `event: new-homepage-modal-click` + `modal-secondary-link: go to current homepage`
  - [x] When a user closes the modal, the modal reports analytics with `event: new-homepage-modal-close`
 
### Accessibility-specific ACs
[See design system](https://design.va.gov/components/modal#accessibility-considerations)

- [x] Modal should have the role=”dialog” attribute.
- [x] Content behind the modal should not scroll while the modal is open.
- [x] The modal and its interactive elements are accessible to all users, regardless of mobile, smartphone or desktop device.
- [x] The modal and its interactive elements are visible onscreen, even if the page is viewed with zoom. 
- [ ] The modal and its interactive elements are accessible to all users, regardless of screenreader, keyboard, voice navigation. or other assistive device.
- [ ] Modal should have an aria-labelledby attribute where the value is the ID of the modal title (you don’t need or want the #).
- [ ] Modal should have an aria-describedby attribute where the value is the ID of a div that contains a description of the modal (you don’t need or want the #).
- [ ] When the modal is closed, focus should go to the skip link at the top of the current/existing or redesigned page, based on user selection. 
- [ ] Pressing the ESC key should close the modal and return the focus to the element that triggered the modal.
- [x] "Try the new homepage" should semantically be a link
- [x] "No Thanks" should semantically be a button
- [ ] Requires accessibility review (Laura Flannery)

### PR DESCRIPTION
**Pull request by jtmst**
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

closes https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11535, https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11976, https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11978

## Summary

- Adds soft launch modal to the homepage behind feature flags on staging/prod

## Related issue(s)
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11535
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11975 - issue found during QA here, non-blocking for this PR
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11976 - issue found during QA here, non-blocking for this PR

## Testing done
Local testing with both staging and prod api data.
QA steps:
- watch vets-website with prod api.
- verify that no modal appears on home, or any other mage
- watch vets-website with staging api
- verify that modal renders as expected
- verify functionality of both buttons, and dismissal on close
- verify modal only appears on homepage

## Screenshots
image

## What areas of the site does it impact?
This uses the announcements framework to display a new modal component on the homepage

## Acceptance criteria

- [ ] Meets issue ACs

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx
Comment (ryguyk) [inline src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx:11]: **SHOULD**
It feels a bit more idiomatic to combine these lines as follows:
```
export function HomepagRedesignModal({ dismiss, vaHomePreviewModal }) {
```

**Review by ryguyk — CHANGES_REQUESTED** [review]

**Review by apisandipas — COMMENTED** [review]
Once small comment that Ryan also called out on the comment above your regex

File: src/platform/site-wide/announcements/config/index.js
Comment (jtmst) [reply src/platform/site-wide/announcements/config/index.js:17]: good catch, change made

Comment (jtmst) [reply src/platform/site-wide/announcements/config/index.js:17]: good catch, change made

**Comment by jilladams** [comment]
@jtmst any way to see the modal on Review Instance?

**Comment by jtmst** [comment]
> @jtmst any way to see the modal on Review Instance?

none of the PR review instances for vets-web seem to be working right now

**Comment by jilladams** [comment]
Oh - I can get to the Review Instance, http://bded6d7898f077e802365f56b610a9cb.review.vetsgov-internal/, just can't see the modal. Wasn't sure if I should be able to or how the flipper works / doesn't in lower envs.

**Comment by jtmst** [comment]
> 

its behind a feature flag on prod api, so it will only show on staging. not sure what environment the review branches point at.

**Comment by ryguyk** [comment]
@jilladams @jtmst The review instances use their own instance of api/Flipper. You can toggle it here:

http://bded6d7898f077e802365f56b610a9cb-api.review.vetsgov-internal/flipper/features

**Review by it-harrison — APPROVED** [review]

**Review by apisandipas — APPROVED** [review]
With previous call outs addressed, this is passable. Be sure to rebase `main` to bring in @ryguyk changes to the config.

**Comment by jilladams** [comment]
PR has file conflicts. 

## QA
* Page load - modal appears
  * **Defect maybe:** x to close loads with a yellow outline. details below
* Page reload - modal appears. Expected, bc session logic work is not yet merged.
* Click not today => modal closes, user lands on VA homepage
* Click "try the new homepage" => Page reload, land on /new-home-page
* Click X to close => modal closes, user lands on VA homepage


Screen Shot 2022-12-14 at 3 28 40 PM
This outline is hard to pinpoint int he dev inspector. It disappears when you click the inspector window, not sure what's revealing it. It shows on page load when modal appears, and disappears if you click on the modal body, too. 


Compared to Find a Form modal, which loads with the heading outlined. (Weird.) This may be an accessibility focus thing? Not sure how to know - happy to talk about it if notes are unclear.
Screen Shot 2022-12-14 at 3 35 56 PM

**Review by jilladams — CHANGES_REQUESTED** [review]
- [ ] Modal should have an aria-describedby attribute where the value is the ID of a div that contains a description of the modal (you don’t need or want the #).
- [x] Willing to approve depending on explanation why the yellow outline on X appears. If it's in fact a defect and requires significant work to fix, we can fast follow in a separate PR in order to get this otherwise merged for QA purposes from the AdHoc team. 

Analytics 
- [x] When a user opts in to the new homepage, the modal reports analytics with event: new-homepage-modal-click + modal-primaryButton-text: try the new homepage
- [x] When a user declines the new homepage, the modal reports analytics with event: new-homepage-modal-click + modal-secondary-link: go to current homepage
- [x] When a user closes the modal, the modal reports analytics with event: new-homepage-modal-close

**Comment by jtmst** [comment]
The outline seems to be the default and desired behavior according to the design team. I was told to ensure that behavior was not changed in this slack thread: https://dsva.slack.com/archives/C01DBGX4P45/p1670874278201309?thread_ts=1670620549.521719&cid=C01DBGX4P45

**Review by ryguyk — CHANGES_REQUESTED** [review]

File: src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx
Comment (jtmst) [reply src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx]: good catch, fixed

**Comment by jtmst** [comment]
> * [ ]  Modal should have an aria-describedby attribute where the value is the ID of a div that contains a description of the modal (you don’t need or want the #).
> * [x]  Willing to approve depending on explanation why the yellow outline on X appears. If it's in fact a defect and requires significant work to fix, we can fast follow in a separate PR in order to get this otherwise merged for QA purposes from the AdHoc team.
> 
> Analytics
> 
> * [x]  When a user opts in to the new homepage, the modal reports analytics with event: new-homepage-modal-click + modal-primaryButton-text: try the new homepage
> * [x]  When a user declines the new homepage, the modal reports analytics with event: new-homepage-modal-click + modal-secondary-link: go to current homepage
> * [x]  When a user closes the modal, the modal reports analytics with event: new-homepage-modal-close

Because we are using the old Modal component (due to the z-index nav bug we discussed) it doesnt look like we are able to overide the default titleid that is used for the aria label and description.  Hopefully we can upgrade to the new component soon and then update.

**Comment by laflannery** [comment]
> Modal should have an aria-labelledby attribute where the value is the ID of the modal title (you don’t need or want the #). = aria-labelledby="modal-announcement-title" @laflannery can you confirm this is correct?

Yes this is correct, please update this (the Find a form modal does this correctly)

> When the modal is closed, focus should go to the skip link at the top of the current/existing or redesigned page, based on user selection. -- not sure how to confirm

This is not happening, this is going the the "Veterans Programs and Services" heading in the footer instead which means if I tab after closing the modal, the next item I go to is the first footer link.  Can this be updated so I am at the top of the page?

>  Pressing the ESC key should close the modal and return the focus to the element that triggered the modal. -- not sure how to confirm

The ESC key does close the modal.  The AC is not entirely accurate as the modal appears immediately upon going to the page.  The focus should work the same as above and take the user to the top of the page

File: src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx
Comment (ryguyk) [inline src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx:11]: **MUST**
This does not need to be exported. More importantly, it _shouldn't_ be exported. The functionality (after being wrapped in a HOC) is being exported as the default at the bottom of the file. See comments on file where this functionality is being imported as well.

File: src/platform/site-wide/announcements/config/index.js
Comment (ryguyk) [inline src/platform/site-wide/announcements/config/index.js:2]: Notice that the typo persists here, but things still work. This is because what's being imported here is the _default_ export from `../components/HomepageRedesignModal`. That is the HOC that is exported at the bottom of the file. On import, you can name it whatever you'd like. As noted in the other comment, we should not export the unwrapped component, because that would allow somebody to import it `import { HomepageRedesignModal } from ...` and that would be the functionality without the connection to the Redux store. Providing the option to import that would be bad.

I do think the typo should be fixed here as well.

**Review by ryguyk — DISMISSED** [review]

File: src/platform/site-wide/announcements/config/index.js
Comment (jtmst) [reply src/platform/site-wide/announcements/config/index.js]: fixed type

File: src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx
Comment (jtmst) [reply src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx]: removed export, thanks!

**Comment by laflannery** [comment]
@jilladams I saw your comment above and wanted to mentioned that 2 is ok, that was the intention

> Skip to content is visible: On modal close with "not today", ESC, or X to close: I see a "Skip to Content" link in header, which disappears if you click anywhere on page.

When the modal is closed (in any fashion) we want the focus to be forced to the top of the page, which is the skip link.  This way, AT users are able to skip to the main body of the content if they choose. Let me know if you have questions on that or if this should be discussed further

**Comment by jilladams** [comment]
@laflannery are we sure we want that to show above the header to all users? It seems buggy for non-screen-reader users.

**Review by jilladams — CHANGES_REQUESTED** [review]

**Comment by laflannery** [comment]
@jilladams Here are my notes and how I tested for each AC. Let me know if there are further questions on this or anything I forgot to put in here that you need.

 - [x] Pressing the ESC key should close the modal and return the focus to the element that triggered the modal. -- not sure how to confirm
- This is not 100% accurate, it should act exactly how closing the modal works using the button or the X.
- This was confirmed using keyboard. When I used ESC, the focus was brought to the skip link and was indicated using the gold outline.
- **Using a screen reader, when I used the ESC key, before the skip link is announced, the first header in the footer (Veterans programs and services) is announced. This is a bug and I am still trying to see if there is a way to resolve this or why it's happening but so far I have no explanation.
- [x]  When the modal is closed, focus should go to the skip link at the top of the current/existing or redesigned page, based on user selection. -- not sure how to confirm
- This was confirmed using keyboard and most. When I used close the modal with either the X or the button, the focus was brought to the skip link and was indicated using the gold outline.
- **Using a screen reader, when I used close the modal with either the X or the button, before the skip link is announced, the first header in the footer (Veterans programs and services) is announced. This is a bug and I am still trying to see if there is a way to resolve this or why it's happening but so far I have no explanation.
- [x]  The modal and its interactive elements are accessible to all users, regardless of mobile, smartphone or desktop device. -- not certain to what extent we should check this at PR stage, vs. AdHoc team doing cross-device testing after merge.
- I have tested this using keyboard, mouse and NVDA are various screen sizes.  As mentioned in the AC it can be further tested on devices once merged as I cannot access it everywhere yet. However, as the code was thoroughly inspected, it should be compatible with various browsers and devices.
- [x]  The modal and its interactive elements are accessible to all users, regardless of screenreader, keyboard, voice navigation. or other assistive device. -- not sure how to check
- - I have tested this using keyboard, mouse and NVDA are various screen sizes.  As mentioned in the AC it can be further tested on devices once merged as I cannot access it everywhere yet. However, as the code was thoroughly inspected, it should be compatible with various browsers and devices.
- [x]  Requires accessibility review (Laura Flannery)

** The bug mentioned I will continue looking at tomorrow and but it is something I have not seen before so it's not clear to me why this is happening.

File: src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx
Comment (jilladams) [inline src/platform/site-wide/announcements/components/HomepageRedesignModal.jsx:44]: Missing `.` at the end of the sentence.

**Review by jilladams — CHANGES_REQUESTED** [review]

**Comment by laflannery** [comment]
I'm not sure if anyone is going to be working on this today with Josh out but I wanted to add my final notes:

1. We got an answer that there does seem to be a way to remove the `aria-label`, so hopefully that works: https://dsva.slack.com/archives/C01DBGX4P45/p1671157719506679?thread_ts=1671127606.381169&cid=C01DBGX4P45
2. Josh sent me a message that he was pretty sure he knew what was going on with the focus SR issue I noted in my last comment and what would need to be looked at to resolve it:

> there was an issue with set-focus not working properly when closing the modal, a solution ive used for similar bugs is adding a 50ms delay before changing focus.  I didnt anticipate it affecting screen readers this way.

3. I apologize but I totally overlooked this yesterday - are we able to change the H3 heading to an H1? The top level heading in any modal should always be an H1. It's not happening here by default because we are not using the standard `modalTitle `attribute and adding our own instead.

**Comment by jilladams** [comment]
re-QA session logic: 
🟢 Refresh page: Modal does not appear
🟢 Open a new browser tab and load http://a109d22ff3eebd5144235d7057fb577f.review.vetsgov-internal/: Modal does not appear
🟢 Quit browser. Open browser. Load http://a109d22ff3eebd5144235d7057fb577f.review.vetsgov-internal/: Modal does not appear

**Review by it-harrison — APPROVED** [review]

**Comment by jilladams** [comment]
Verified: 

1. Secondary button
2. Period at end of copy
3. H1

@apisandipas @jtmst only issues left here: 
1. @ryguyk  re-review or override 
2. Laura's last note: re focus SR issue -- has that been addressed? If so, let Laura know to verify.

**Comment by jtmst** [comment]
> Verified:
> 
> 1. Secondary button
> 2. Period at end of copy
> 3. H1
> 
> @apisandipas @jtmst only issues left here:
> 
> 1. @ryguyk  re-review or override
> 2. Laura's last note: re focus SR issue -- has that been addressed? If so, let Laura know to verify.

Laura gave ok to launch to staging, and address the SR issue separately,  looks like Bryan addressed everything else already. Just need review approval

**Review by jilladams — APPROVED** [review]

**Comment by jilladams** [comment]
@jtmst I think you've addressed all Ryan's feedback as well. If that's the case, and Bryan has signed off, sounds like Ryan won't have time for PRs today, so if you are otherwise ready, i think you can override the blocker on his review. (I think that's possible?)

**Comment by apisandipas** [comment]
I've pushed a commit to this branch that fixes the issue with the modal that was identified in:
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11978

It should now be resolved on the review instance associated with this PR. This was done, one, in the interest of getting these deployed together, since they would have been merged at the same time anyways and also to resolve some local git issues we were experiencing on Josh's machine.

**Comment by jilladams** [comment]
(RI is 502ing. May come up after a bit)

**Pull request merged**
